Kill globals and generalize Backend traits

clux opened the pull request:
This is meant to complete #7 once it's done.

clux committed:
rename trait method get_tarball_url to get_component_info

clux committed:
generalize Backend trait method names

clux committed:
s/tarball/location in Component struct

clux committed:
remove global buckets and get everything building without it

leaves a couple of TODOs here and there (particularly for `update`), but a lot of it just disappeared cleanly by itself.

clux committed:
remove useless function

clux committed:
fix update to look at intersection

This actually fixes a bunch of things, as well as clarifies trait roles
and output log for update.

- supportedEnvironments added to manifest (and its init)
- Backend.get_versions() now returns in descending order **BREAKS lal
query**
- get_latest_supported_versions auto-implemented on CachedBackend based
off get_versions()

Now if you call lal update in a place that supports two environments you
may see this:

lal::update: Fetch artful libopus
lal::storage::download: Last versions for libopus in xenial env is {12, 13, 16}
lal::storage::download: Last versions for libopus in artful env is {16}
lal::update: Fetch artful libopus=16
lal::update: Upgrading libopus from 13 to 16

clux commented:
Got a few breaking changes now:

- `lal query` lists versions in reverse direction now (so people using `| tail` will be screwed)
- manifest.json now requires an extra `supportedEnvironments` key in each repo
- `lal query` without an `-e env` flag will crash (TODO: want to handle this better)
- `lal export` without an `-e env` flag will crash (TODO: want to handle this better)

But beyond that, things work as before, just with some more logging in `update`

bencord0 commented on src/export.rs:
```diff
@@ -9,14 +9,14 @@ pub fn export<T: CachedBackend + ?Sized>(
 
     let dir = output.unwrap_or(".");
-    info!("Export {} {} to {}", env.unwrap_or("global"), comp, dir);
+    info!("Export {} {} to {}", env, comp, dir);
```
Yeah, this is much nicer now.

bencord0 commented on src/main.rs:
```diff
@@ -39,14 +39,14 @@ fn handle_manifest_agnostic_cmds(
                     a.value_of("component").unwrap(),
                     a.value_of("output"),
-                    explicit_env)
+                    explicit_env.unwrap()) // TODO: enforce unwrap condition
```
Do you want to handle these `unwrap()` now? or clean them up in another PR?

bencord0 commented on src/storage/download.rs:
```diff
@@ -94,21 +93,41 @@ impl<T: ?Sized> CachedBackend for T
 where
     T: Backend,
 {
+    /// Get the latest version of a component across all supported environments
```
Add the note from #7 about how this returns a version which is available in all (supported) environments.
Even though a higher version is available, if it isn't everywhere, it won't be chosen.

bencord0 commented on src/storage/traits.rs:
```diff
@@ -38,25 +40,23 @@ pub struct Component {
+    /// If no version is given, figure out what latest is
+    fn get_component_info(&self, name: &str, ver: Option<u32>, loc: &str) -> LalResult<Component>;
 
     /// Publish a release build (ARTIFACT dir) to a specific location
```
`/// Publish a release build's ARTIFACT to a backend specific location.`
If we're renaming the method, maybe just `publish_artifact`, with a note that this means everything in the directory, the tarball and the lockfile (and maybe a signature when we eventually support those).

bencord0 commented on src/update.rs:
```diff
@@ -50,7 +50,19 @@ pub fn update<T: CachedBackend + ?Sized>(
+                None => {
+                    return Err(CliError::NoIntersectedVersion(comp.clone()));
+                }
+            };
```
Sometimes I just enjoy reading rust error handlers.
It's something about knowing that after this line, `ver` is something sensible.

clux replied:
If you liked that, replace that entire match with a chained on:

```sh
.ok_or(CliError::NoIntersectedVersion(comp.clone()))?
```

bencord0 replied:
... tbh, I prefer it the other way.

clux replied:
hah, aw

clux committed:
rename a publish method, clean up error handling in update + cargo fmt

thought i could run cargo fmt on just a file, but it ran on all the
things, so have to add it in this commit. :/

clux committed:
remove an old error that's no longer necessary to restrict for

clux committed:
bump clap

clux committed:
make missing env in query + export explicit

mostly because clap crashes if we try to require environment in the arg
parser at the moment.

clux committed:
remove more references to global dependencies

bencord0 commented on src/main.rs:
```diff
@@ -39,10 +39,10 @@ fn handle_manifest_agnostic_cmds(
                     a.value_of("component").unwrap(),
                     a.value_of("output"),
-                    explicit_env.unwrap()) // TODO: enforce unwrap condition
+                    explicit_env)
```
cool

bencord0 commented on SPEC.md:
```diff
@@ -27,6 +27,7 @@ A per-repo file. Format looks like this (here annotated with illegal comments):
 {
   "name": "libwebsockets",  // name of repo
   "environment": "centos",  // name of environment found in the lal config
+  "supportedEnvironments": ["centos", "xenial"],
```
w00t
